crates/subspace-networking/src/protocols/request_response/handlers/cached_piece_by_index.rs

```diff
@@ -0,0 +1,128 @@
+//! Helper for incoming cached piece requests.
+//!
+//! Request handler can be created with [`CachedPieceByIndexRequestHandler`].
+
+use crate::protocols::request_response::handlers::generic_request_handler::{
+    GenericRequest, GenericRequestHandler,
+};
+use derive_more::{Deref, DerefMut, From, Into};
+use libp2p::kad::K_VALUE;
+use libp2p::multiaddr::Protocol;
+use libp2p::{Multiaddr, PeerId};
+use multihash::Multihash;
+use parity_scale_codec::{Compact, CompactLen, Decode, Encode, EncodeLike, Input, Output};
+use subspace_core_primitives::pieces::{Piece, PieceIndex};
+
+/// Cached-piece-by-index request.
+///
+/// This is similar to `PieceByIndexRequest`, but will only respond with cached pieces.
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct CachedPieceByIndexRequest {
+    /// Request key - piece index
+    pub piece_index: PieceIndex,
+    /// Additional pieces that requester is interested in if they are cached locally
+    pub cached_pieces: Vec<PieceIndex>,
+}
+
+impl GenericRequest for CachedPieceByIndexRequest {
+    const PROTOCOL_NAME: &'static str = "/subspace/cached-piece-by-index/0.1.0";
+    const LOG_TARGET: &'static str = "cached-piece-by-index-request-response-handler";
+    type Response = CachedPieceByIndexResponse;
+}
+
+/// Closest peers
+#[derive(Debug, Default, PartialEq, Eq, Clone, From, Into, Deref, DerefMut)]
+pub struct ClosestPeers(Vec<(PeerId, Vec<Multiaddr>)>);
+
+impl Encode for ClosestPeers {
+    fn size_hint(&self) -> usize {
+        let mut size = Compact::compact_len(&(self.0.len() as u32));
+
+        for (peer_id, addresses) in &self.0 {
+            size += peer_id.as_ref().encoded_size();
+            size += Compact::compact_len(&(addresses.len() as u32));
+
+            for address in addresses {
+                size += address.as_ref().encoded_size();
+            }
+        }
+
+        size
+    }
+
+    fn encode_to<T: Output + ?Sized>(&self, dest: &mut T) {
+        Compact::from(self.0.len() as u32).encode_to(dest);
+
+        for (peer_id, addresses) in &self.0 {
+            peer_id.as_ref().encode_to(dest);
+            Compact::from(addresses.len() as u32).encode_to(dest);
+
+            for address in addresses {
+                address.as_ref().encode_to(dest);
+            }
+        }
+    }
+}
+
+impl EncodeLike for ClosestPeers {}
+
+impl Decode for ClosestPeers {
```

Comment:
Nontrivial decoding: at least simple "encode-decode-compare" test would be great.